sdk/storage/Azure.Storage.Common/src/StorageRequestFailedException.cs

```diff
@@ -49,50 +63,107 @@ public StorageRequestFailedException(Response response, string message = null)
-            => ResponseExceptionExtensions.CreateRequestFailedMessageAsync(message, response, false).GetAwaiter().GetResult();
+        /// <param name="message">The default message.</param>
+        /// <param name="response">The error response.</param>
+        /// <param name="errorCode">An optional error code.</param>
+        /// <param name="additionalInfo">Optional additional information.</param>
+        /// <returns>The exception's Message.</returns>
+        private static string CreateMessage(
+            Response response,
+            string message,
+            string errorCode,
+            IDictionary<string, string> additionalInfo)
+        {
+            // Start with the message, status, and reason
+            var messageBuilder = new StringBuilder()
+                .AppendLine(message)
+                .Append("Status: ")
+                .Append(response.Status.ToString(CultureInfo.InvariantCulture))
+                .Append(" (")
+                .Append(response.ReasonPhrase)
+                .AppendLine(")");
 
-        /// <summary>
-        /// Gets the x-ms-request-id header that uniquely identifies the
-        /// request that was made and can be used for troubleshooting.
-        /// </summary>
-        /// <param name="response">Response of the failed request.</param>
-        /// <returns>
-        /// The x-ms-request-id header that uniquely identifies the request
-        /// that was made and can be used for troubleshooting.
-        /// </returns>
-        private static string GetRequestId(Response response)
-            => response.Headers.TryGetValue("x-ms-request-id", out var value) ? value : null;
+            // Make the Storage ErrorCode especially prominent
+            if (!String.IsNullOrEmpty(errorCode) ||
+                response.Headers.TryGetValue(ErrorCodeHeaderName, out errorCode))
+            {
+                messageBuilder
+                    .AppendLine()
+                    .Append("ErrorCode: ")
+                    .AppendLine(errorCode);
+            }
+
+            // A Storage error's Content is (currently) always the ErrorCode and
+            // AdditionalInfo, so we skip the specific Content section
+            if (additionalInfo != null && additionalInfo.Count > 0)
+            {
+                messageBuilder
+                    .AppendLine()
+                    .AppendLine("Additional Information:");
+                foreach (var info in additionalInfo)
+                {
+                    messageBuilder
+                        .Append(info.Key)
+                        .Append(": ")
+                        .AppendLine(info.Value);
+                }
+            }
+
+            // Include the response headers
+            messageBuilder
+                .AppendLine()
+                .AppendLine("Headers:");
+            foreach (var responseHeader in response.Headers)
+            {
+                messageBuilder
```

Comment:
Are there any headers we don't want to put in here?

Reply:
This is a great question.  No, we're okay.  These are only response headers and the only `x-ms-copy-source` response headers are returned on successful operations.

There's an ongoing discussion about whether we want to support logging redaction _(I think we do)_ and we should plumb that here and the base `RequestFailedException` when it's implemented.  /cc @KrzysztofCwalina